CAL-346 Added documentation for NITF mappings

emmberk opened the pull request:
#### What does this PR do?
This PR adds documentation for NITF mappings.

**NOTE**: Any concerns about corrections to current mappings should be addressed in new Alliance tickets.
#### Who is reviewing it? 
@lcrosenbu @peterhuffer @AzGoalie 
#### Select relevant component teams:
@codice/docs @ricklarsen 
@codice/data
#### Choose 2 committers to review/merge the PR.
@bdeining
@clockard
#### How should this be tested?

- Build docs.
- Review new content and check that it matches the `imaging-transformer-nitf` code.
- Check for typos.
- Confirm formatting (especially new lines in the `Translation` column) looks correct in the pdf and the html Documentation ([attached](https://github.com/codice/alliance/files/1345101/documentation.zip)).


#### Any background context you want to provide?
#### What are the relevant tickets?
[CAL-346](https://codice.atlassian.net/browse/CAL-346)
#### Screenshots
![image](https://user-images.githubusercontent.com/8041246/31027068-b97efdac-a4fd-11e7-91df-f190249cb4bd.png)
#### Checklist:
- [x] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

emmberk commented on catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/StdidcAttribute.java:
```diff
@@ -103,6 +101,22 @@
+          COUNTRY_SHORT_NAME,
+          tre -> getAlpha3CountryCode(TreUtility.convertToString(tre, COUNTRY_SHORT_NAME)),
+          new LocationAttributes().getAttributeDescriptor(Location.COUNTRY_CODE),
+          "");
```
This attribute was updated to use the `Normalized attribute` syntax that is used in other `xxxAttribute` classes, like https://github.com/codice/alliance/blob/1faf1f4dd3bb7262cc4088dc74b6b63b7b059519/catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/NitfHeaderAttribute.java#L165-L171.

One of the benefits, besides consistency, is to use the already-existing `location.country-code` `AttributeDescriptor` instead of defining it in this class. The corresponding constructor was also added.

emmberk commented on catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/ImageAttribute.java:
```diff
@@ -149,10 +151,17 @@
+          "IID2",
+          ImageSegment::getImageIdentifier2,
+          new IsrAttributes().getAttributeDescriptor(Isr.MISSION_ID),
+          "");
```
This `Attribute` was moved here from https://github.com/codice/alliance/blob/1faf1f4dd3bb7262cc4088dc74b6b63b7b059519/catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfImageTransformer.java#L157-L167.

emmberk replied:
@peterhuffer, mostly formatting and typos. @clockard did the initial draft of the mappings, and I double-checked everything as I was working on this PR, so the content of the mappings has already been reviewed by me.

A few other things to review:

- Many rows repeat the same NITF Field because it is mapped to multiple attributes. Does this make sense? Should this be reformatted?
- Check the `Translation` column. Do the notes make sense?
- Does the separation of tables and throwing them all in the appendix make sense? Should I reorganize the rows?
- Is there missing `warning` text or notes that should be added to these tables to provide some context?

emmberk commented on catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/NitfImageTransformer.java:
```diff
@@ -152,21 +151,6 @@ protected void handleGeometry(ImageSegment imageSegmentHeader, List<Polygon> pol
-      LOGGER.debug("Setting the metacard attribute [{}, {}]", Isr.MISSION_ID, missionId);
-      metacard.setAttribute(new AttributeImpl(Isr.MISSION_ID, missionId));
-    }
-  }
```
This `Attribute` was moved to https://github.com/codice/alliance/blob/e7c036fa8a556d1175a55e69bcba6b64ca4ecbae/catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/ImageAttribute.java.

emmberk commented on catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/TextAttribute.java:
```diff
@@ -85,7 +90,7 @@
   TEXT_CLASSIFICATION_AUTHORITY_TYPE(
       "text-classification-authority-type",
-      "TSCA TP",
+      "TSCATP",
```
It doesn't look like this should have a space:
http://www.gwg.nga.mil/ntb/baseline/docs/stdi0006/NCDRD_20Dec06.pdf.

emmberk commented on catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/ImageInputTransformerTest.java:
```diff
@@ -145,27 +145,6 @@ public void testNitfParsing() throws Exception {
-    Metacard metacard = metacardFactory.createMetacard("noIdentifierTest");
-    transformer.handleMissionIdentifier(metacard, null);
-    assertThat(metacard.getAttribute(Isr.MISSION_ID), nullValue());
-  }
```
These tests are no longer needed now that the mission.id `Attribute` was moved to https://github.com/codice/alliance/blob/e7c036fa8a556d1175a55e69bcba6b64ca4ecbae/catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/ImageAttribute.java.

peterhuffer replied:
❓ There is an `ImageAttributeTest.java` class, should we add a test for the MissionID in there?

emmberk commented on distribution/docs/src/main/resources/content/_metadata-attributes/NITF-Attribute-Mappings-table-contents.adoc:
```diff
@@ -0,0 +1,38 @@
+:status: published
+:summary: NITF Attribute Mappings.
+
+.[[NITF_ACFTB_Attribute_Mappings]]NITF Attribute Mappings
```
I plan to write a new ticket for reworking this table and correcting the mappings. I will link it here.

emmberk commented on catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/image/ImageAttribute.java:
```diff
@@ -149,10 +151,17 @@
+          "IID2",
+          ImageSegment::getImageIdentifier2,
+          new IsrAttributes().getAttributeDescriptor(Isr.MISSION_ID),
+          "");
```
@peterhuffer, mostly formatting and typos. @clockard did the initial draft of the mappings, and I double-checked everything as I was working on this PR, so the content of the mappings has already been reviewed by me.

A few other things to review:

- Many rows repeat the same NITF Field because it is mapped to multiple attributes. Does this make sense? Should this be reformatted?
- Check the `Translation` column. Do the notes make sense?
- Does the separation of tables and throwing them all in the appendix make sense? Should I reorganize the rows?
- Is there missing `warning` text or notes that should be added to these tables to provide some context?

emmberk commented on catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/image/ImageInputTransformerTest.java:
```diff
@@ -686,7 +668,9 @@ private InputStream getInputStream(String filename) {
-    map.put(ImageAttribute.IMAGE_IDENTIFIER_2_ATTRIBUTE, "- BASE IMAGE -");
+    final String imageIdentifier2 = "- BASE IMAGE -";
+    map.put(ImageAttribute.MISSION_ID_ATTRIBUTE, imageIdentifier2);
+    map.put(ImageAttribute.IMAGE_IDENTIFIER_2_ATTRIBUTE, imageIdentifier2);
```
@peterhuffer, following the convention for the other attributes in the `ImageAttribute` class, I added this line to make sure that the `isr.mission-id` attribute is still unit tested even though it was moved.

ahoffer commented:
Reviewed rendered documentation changes. Emily provided HTML and PDF files for review. The table formatting looks good in both versions.

The empty cells in the translation column are ambiguous. Suggest putting in "None" or "No translation" or "N/A" or something in the empty cells.

"substring(7,11)" can be mean different things if the bounds are exclusive or inclusive. Might be clearer to say "last four characters"

3.4028235e+38f is an odd number for depth-of-snow. Maybe say "MAX FLOAT" or something.

ahoffer commented:
build now

emmberk commented:
I addressed @ahoffer's suggestions:
- Clarified undefined
- Added quotes around "Unknown"
- Added note aboout unable to parse
- Replaced max floats
- Removed substring translation
- Added none in blank cells

emmberk commented:
build now

emmberk commented:
build now
